pkg/controller/gpudevice/vgpu_controller.go

```diff
@@ -0,0 +1,357 @@
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"reflect"
+	"sync"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+
+	"github.com/harvester/pcidevices/pkg/apis/devices.harvesterhci.io/v1beta1"
+	"github.com/harvester/pcidevices/pkg/deviceplugins"
+	"github.com/harvester/pcidevices/pkg/util/gpuhelper"
+)
+
+var (
+	pluginLock sync.Mutex
+)
+
+const (
+	DefaultNS  = "harvester-system"
+	KubevirtCR = "kubevirt"
+)
+
+func (h *Handler) OnVGPUChange(_ string, vgpu *v1beta1.VGPUDevice) (*v1beta1.VGPUDevice, error) {
+	if vgpu == nil || vgpu.DeletionTimestamp != nil || vgpu.Spec.NodeName != h.nodeName {
+		return vgpu, nil
+	}
+
+	discoveredVGPUStatus, err := gpuhelper.FetchVGPUStatus(v1beta1.MdevRoot, v1beta1.SysDevRoot, v1beta1.MdevBusClassRoot, vgpu.Spec.Address)
+	if err != nil {
+		return vgpu, fmt.Errorf("error generating vgpu %s status: %v", vgpu.Name, err)
+	}
+
+	// gpu spec is enabled and discovered status indicates no configuration
+	if vgpu.Spec.Enabled && discoveredVGPUStatus.VGPUStatus == v1beta1.VGPUDisabled {
+		return h.enableVGPU(vgpu)
+	}
+
+	if !vgpu.Spec.Enabled && discoveredVGPUStatus.VGPUStatus == v1beta1.VGPUEnabled {
+		return h.disableVGPU(vgpu)
+	}
+	// perform enable disable operation //
+	if !reflect.DeepEqual(discoveredVGPUStatus, vgpu.Status) {
+		vgpu.Status = *discoveredVGPUStatus
+		return h.vGPUClient.UpdateStatus(vgpu)
+	}
+
+	return nil, nil
+}
+
+func (h *Handler) SetupVGPUDevices() error {
+	vGPUDevices, err := gpuhelper.IdentifyVGPU(h.options, h.nodeName)
+	if err != nil {
+		return fmt.Errorf("error identifying vgpu devices: %v", err)
+	}
+	return h.reconcileVGPUSetup(vGPUDevices)
+}
+
+func (h *Handler) reconcileVGPUSetup(vGPUDevices []*v1beta1.VGPUDevice) error {
+	set := map[string]string{
+		v1beta1.NodeKeyName: h.nodeName,
+	}
+
+	vGPUList, err := h.vGPUCache.List(labels.SelectorFromSet(set))
+	if err != nil {
+		return err
+	}
+
+	for _, v := range vGPUDevices {
+		existingVGPU := containsVGPU(v, vGPUList)
+		if existingVGPU != nil {
+			if !reflect.DeepEqual(v.Status, existingVGPU.Status) {
+				// on reboot the vGPU status will not match the state in CRD
+				// in which case we should if needed reset the vGPU status and
+				// allow reconcile to flow through
+				existingVGPU.Status = v.Status
+				if _, err := h.vGPUClient.UpdateStatus(existingVGPU); err != nil {
+					return err
+				}
+			}
+		} else {
+			if _, err := h.vGPUClient.Create(v); err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, v := range vGPUList {
+		parentDeviceEnabled, err := h.isParentGPUEnabled(v.Spec.ParentGPUDeviceAddress)
+		if err != nil {
+			return err
+		}
+
+		// if parentDevice is enabled, then skip deletion
+		// as controller will reconcile and re-configure vGPU when device parent SRIOVGPU device
+		// is reconfigured post reboot
```

Comment:
Can you help to explain how SRIOVGPU trigger the reconcile on VGPU? Through the node controller's [periodic reconciliation every 30 seconds](https://github.com/ibrokethecloud/pcidevices/blob/55de4d75a7cd9fac885bbad31306ff83e0808bd7/pkg/controller/nodes/node_controller.go#L123)?

Reply:
the node object is a fake crd with only one purpose which is to regularly reconcile hardware objects to detect changes in underlying OS. In some cases like sriovnetworkdevices and gpu's if you passthrough the device directly to a guest then we should ignore it from usage in srviogpudevices and subsequent vgpudevices of Network VF's as the device is no longer being managed the host. The reverse is also true, if the GPU or NIC is used for VF's then PCIDevice object for this should not exist to avoid the PF being passed through to a guest.